src/resolve/resolved-varlink.c

```diff
@@ -830,7 +830,7 @@ static void resolve_service_all_complete(DnsQuery *query) {
                 goto finish;
 
         if (isempty(type)) {
-                r = sd_varlink_error(q->varlink_request, "io.systemd.Resolve.ServiceNotProvided", NULL);
+                r = sd_varlink_error(q->varlink_request, "io.systemd.Resolve.InconsistentServiceRecords", NULL);
```

Comment:
Given that https://github.com/systemd/systemd/issues/33947 can't be fixed I think another option would be to actually resolve those things and document that it's possible that `.canonical` doesn't make sense at all. It should make this API somewhat consistent. I took a look at what `mDNSResponder` does and as far as I can see it resolves those services (though they can't be browsed and `null`s aren't returned anywhere)

Reply:
Even though #33947 may be important, but it should be irrelevant to this PR, at least.

Reply:
I don't think that issue is particularly important but it's basically the same issue where CNAMEs point to nonsensical names. If resolved can't reject those service names it probably shouldn't try to enforce those rules here too. It's just inconsistent.

Reply:
I mean, even though the root cause may be equivalent or relevant, this is superficially about error code, and not about the logic of validation.
Your point should be discussed in #33947.

Reply:
The point I'm trying to make is that mDNSResponder (which is the reference implementation of mDNS) resolves those services without producing any errors and resolved should probably does that too (so https://github.com/systemd/systemd/commit/b48ab08732a76b7337628e1e716f11c687000903 should probably be reverted). The `.canonical` field can be nonsensical and it should probably be documented.

Reply:
I am not sure we should follow the behavior. But, anyway, that's out of scope of this PR, and should be commented in #33947 (yeah, I know you already commented there).

Reply:
Dunno. I still think it's inconsistent but I don't feel strongly about this.

Other than that I can't seem to reproduce https://github.com/systemd/systemd/issues/33947 anymore. It seems it fails to follow CNAMEs and can't return some RRs from its cache properly. I'm not sure where it broke this time.
```sh
$ resolvectl show-cache
...
Scope protocol=mdns family=AF_INET ifindex=2 ifname=eth0
_qotd._tcp.local IN SRV 0 0 123 host.local
host.local IN AAAA ::
beep._qotd._tcp.local IN CNAME _qotd._tcp.local
_qotd._tcp.local IN TXT ""
host.local IN A 192.168.56.3

$ resolvectl query --type SRV beep._qotd._tcp.local
beep._qotd._tcp.local: resolve call failed: Name '_qotd._tcp.local' does not have any RR of the requested type

$ resolvectl query --type CNAME beep._qotd._tcp.local
beep._qotd._tcp.local IN CNAME _qotd._tcp.local             -- link: eth0

-- Information acquired via protocol mDNS/IPv4 in 50.9ms.
-- Data is authenticated: no; Data was acquired via local or encrypted transport: no
-- Data from: cache

$ resolvectl query --type SRV _qotd._tcp.local 
_qotd._tcp.local: resolve call failed: Name '_qotd._tcp.local' does not have any RR of the requested type

$ resolvectl query --type A host.local
host.local IN A 192.168.56.3                                -- link: eth0
-- Information acquired via protocol mDNS/IPv4 in 44.7ms.
-- Data is authenticated: no; Data was acquired via local or encrypted transport: no
-- Data from: cache

```

Reply:
Having taken a look at that closer a `_qotd._tcp` service was advertised on that machine by systemd-resolved so apparently with `--zone=no/SD_RESOLVED_NO_ZONE` it can extract those RRs from its cache and that issue is still reproducible but I'll leave it to someone else to figure out how it all should work together.

Reply:
Anyway, please comment on #33947. Please...

Reply:
@yuwata https://github.com/systemd/systemd/issues/33947 is a separate issue. I mentioned it here to show why I think that this API is inconsistent. In the context of this PR https://github.com/systemd/systemd/commit/b48ab08732a76b7337628e1e716f11c687000903 was relevant and the point I was trying to make is that that commit should have been reverted instead of returning error codes. It has been decided that it shouldn't be reverted and that makes sense too.

Reply:
I do not think the commit should be reverted. I do not think the reference implementation handles broken upstream entries correctly. In such case, the table is completely borked, and we should really provide some recognizable error code.

In the same sense, I think that we should return an error in the case #33947, as I already commented https://github.com/systemd/systemd/pull/34105#discussion_r1741334434.

Reply:
> In the same sense, I think that we should return an error in the case https://github.com/systemd/systemd/issues/33947, as I already commented https://github.com/systemd/systemd/pull/34105#discussion_r1741334434

I agree that ideally it's what resolved should do to be consistent but the problem is that it can't do that given that the API tries to do several things and can't tell them apart to enforce the rules properly.

Either way the fact that its own services can prevent it from following CNAMEs and extracting RRs from the cache described in https://github.com/systemd/systemd/pull/34105#discussion_r1742107235 seems worse to me but I'm not going to pursue that.